src/libs/elektra/key.c

```diff
@@ -244,218 +244,220 @@ Key * keyVNew (const char * name, va_list va)
- * wrong metadata and is not able to copy empty or cascading names:
- *
- * @snippet keyCopy.c Individual Copy
+ * Calling `keyCopy (dest, NULL, KEY_CP_ALL)` is different from calling keyClear().
+ * The key will not be fully reset, the reference counter and internal flags
+ * will remain unchanged. Additionally, keyCopy() respects keyLock() state,
+ * while keyClear() always works.
  *
+ * @snippet keyCopy.c Clear
  *
  * @param dest the key which will be written to
  * @param source the key which should be copied
- *     or NULL to clean the destination key
+ *     or NULL to clear the data of @p dest
+ * @param flags specifies which parts of the key should be copied
  * @ingroup key
- * @retval -1 on failure when a NULL pointer
- *     was passed for dest or a dynamic property could not
- *     be written. The content will be unmodified then.
- * @retval 0 when dest was cleaned
- * @retval 1 when source was successfully copied
- * @see keyDup() to get a duplication of a key
+ *
+ * @return @p dest or NULL, in case of error. Possible errors are:
+ *   - memory allocation problems
+ *   - a part of @p dest that should be modified (e.g. name, value) was marked read-only,
+ *     e.g. the name of @p dest will be read-only if @p dest is part of a KeySet
+ *   - @p dest is NULL
+ *   - both #KEY_CP_VALUE and #KEY_CP_STRING are set in @p flags
+ *   - both #KEY_CP_STRING is set in @p flags and @p source is a binary key (keyIsBinary())
  */
-int keyCopy (Key * dest, const Key * source)
+Key * keyCopy (Key * dest, const Key * source, elektraCopyFlags flags)
 {
-	if (!dest) return -1;
+	if (dest == NULL) return NULL;
 
-	if (test_bit (dest->flags, KEY_FLAG_RO_NAME) || test_bit (dest->flags, KEY_FLAG_RO_VALUE) ||
-	    test_bit (dest->flags, KEY_FLAG_RO_META))
-	{
-		return -1;
-	}
+	if (test_bit (dest->flags, KEY_FLAG_RO_NAME) && test_bit (flags, KEY_CP_NAME)) return NULL;
+	if (test_bit (dest->flags, KEY_FLAG_RO_VALUE) && test_bit (flags, KEY_CP_VALUE)) return NULL;
+	if (test_bit (dest->flags, KEY_FLAG_RO_META) && test_bit (flags, KEY_CP_META)) return NULL;
 
-	if (!source)
+	if (test_bit (flags, KEY_CP_STRING) && test_bit (flags, KEY_CP_VALUE)) return NULL;
+
+	if (source == NULL)
 	{
-		keyClear (dest);
-		return 0;
+		if (test_bit (flags, KEY_CP_NAME))
+		{
+			keySetName (dest, "/");
+		}
+		if (test_bit (flags, KEY_CP_VALUE))
+		{
+			keySetRaw (dest, NULL, 0);
+		}
+		if (test_bit (flags, KEY_CP_META))
+		{
+			ksClear (dest->meta);
+		}
+		return dest;
 	}
 
-	// remember dynamic memory to be removed
-	char * destKey = dest->key;
-	char * destUKey = dest->ukey;
-	void * destData = dest->data.c;
-	KeySet * destMeta = dest->meta;
+	if (test_bit (flags, KEY_CP_STRING) && keyIsBinary (source)) return NULL;
 
-	// duplicate dynamic properties
-	if (source->key)
-	{
-		dest->key = elektraStrNDup (source->key, source->keySize);
-		if (!dest->key) goto memerror;
-	}
-	else
-	{
-		dest->key = 0;
-	}
+	if (source == dest) return dest;
```

Comment:
Sounds like this basic check could be done earlier before we do the many other checks at all.

Reply:
Moving this further up changes the behaviour. We need to decide, if `source == dest` is a special case that works, or if the same constraints apply (e.g. `keyCopy` fails if `dest` is read-only).